### PR TITLE
journal: Add option to skip boot kmsg events

### PR DESCRIPTION
**Pull request by ahkok**
The `BootKMsg` boolean is added. It defaults to `true` in the default
system namespace, and doesn't change the default behavior of the
journal.

If set to `false`, then, when `systemd-journald` starts in the default
namespace, it will ignore any existing message in the `kmsg` ring buffer
and not insert these into the journal. We do this to reduce the time
needed for the journal to start and reduce CPU overhead, since the
kernel ring buffer can be large, and contain information that the user
perhaps does not care about (e.g. initcall debug output).

The information remains accessible through `dmesg` in case it needs
to be retrieved later. This contains code originally added to
ClearLinux by Dimitri -xnox- Ledkov in 2015.

### CODE REVIEW

**Comment by ahkok** [comment]
Cc: @xnox

File: src/journal/journald-kmsg.c [hunk 1]
```diff
@@ -384,6 +384,17 @@ int server_open_dev_kmsg(Server *s) {
...
+        if (!s->boot_kmsg) {
+                /* clear out /dev/kmsg, we don't want all its messages */
+                char buffer[40960];
+                while (1) {
+                        int ret;
+                        ret = read(s->dev_kmsg_fd, buffer, 40960);
+                        if (ret <= 0)
+                                break;
+                }
+        }
```
Comment (poettering) [inline src/journal/journald-kmsg.c:396]: we have flush_fd() for this kind of stuff, which does the same

Comment (poettering) [reply src/journal/journald-kmsg.c:396]: this doesn't look right in general, i.e. when journald is restarted we would run this and drop everything happening during the restart. 

The kernel has a seqnum concept which allows us to identify which records we already have seen and which ones not. if the feature you are looking for is indeed desirable it should be built on that: see journald-kmsg.c, in dev_kmsg_record() and  server_open_kernel_seqnum(), where we compare and manage seqnums. We store the last msg seqnum we saw in a file /run/kernel-seqnum. Most likely it suffices to initialize that one after the first call to server_flush_dev_kmsg() and leave it unset before that.

File: src/journal/journald-gperf.gperf [hunk 2]
```diff
@@ -22,6 +22,7 @@ Journal.Storage,            config_parse_storage,    0, offsetof(Server, storage
 Journal.Compress,           config_parse_compress,   0, offsetof(Server, compress)
 Journal.Seal,               config_parse_bool,       0, offsetof(Server, seal)
 Journal.ReadKMsg,           config_parse_bool,       0, offsetof(Server, read_kmsg)
+Journal.BootKMsg,           config_parse_bool,       0, offsetof(Server, boot_kmsg)
```
Comment (poettering) [inline src/journal/journald-gperf.gperf:25]: this shouldn't be a new option, but maybe a new enum value of ReadKMsg=. But I am not sure this is a really convincing feature. I mean, i am not sure it's a good option to just waste the early boot msgs entirely. It's a cheap hack, cutting corners, not an actual fix for possible performance issues...

**Comment by xnox** [comment]
xnox from 2020, doesn't think it is a good idea to drop kmsg on the floor. I thought the optimization that was desired is to have journald started earlier and start accepting messages / let systemd start units ordered after journald, whilst kmsg are not yet read and stored in journal.

I haven't touched journald in years, but i think we can now fork a process / schedule async task to consume kmsg, whilst declaring journald started. And hopefully not loose any kmsg.

File: src/journal/journald-kmsg.c (hunk 1 above)
Comment (ahkok) [reply src/journal/journald-kmsg.c:396]: Thanks, I'll look into that

**Comment by ahkok** [comment]
We carry 2 patches on the journal - one indeed swaps "flush / start" around so that the journal can accepting messages earlier. I didn't include that patch as part of this PR and we've always kept that separately.

This is purely about reducing CPU usage. ClearLinux enables `initcall_debug` at boot time and generates a lot of kmsg data, all that needs to be processed and stored in the journal without that patch, but there's very little compelling reasons to delay boot by 50ms-200ms (easily, let's be realistic) if people can just use `sudo dmesg` to get what they need anyway. In 5+ years time, we had nobody ask about this...so my conclusion is that this is still worth doing.

I was certainly contemplating keeping this out of upstream, and that obviously remains an option (we carry about 40 or so patches against systemd in ClearLinux), and rebasing those isn't a big deal, but once in a while I feel that patches are going a little further and I think more people may be interested in the underlying intent of the patch - hack or not - my intent isn't to "drop the hack" but to improve on this, whether we upstream it or not.

Anyway, for normal users this should be a no-op, it doesn't drop anything and behavior will remain the same as is (that's the goal) so that in ClearLinux we can "flip" the default the other way, but upstream has this option for users who want to shave off some CPU at boot time.

**Comment by ahkok** [comment]
I'll take some time to ingest the feedback - thanks @xnox for not hiding on me ;)

**Comment by poettering** [comment]
> This is purely about reducing CPU usage. ClearLinux enables `initcall_debug` at boot time and generates a lot of kmsg data, all that needs to be processed and stored in the journal without that patch, but there's very little compelling reasons to delay boot by 50ms-200ms (easily, let's be realistic) if people can just use `sudo dmesg` to get what they need anyway. In 5+ years time, we had nobody ask about this...so my conclusion is that this is still worth doing.

if this is about debugging, then maybe beef up ReadKMsg= to optionally filter by log level? let's say you could write ReadKMsg=info then this could mean read everything from kmsg with log level LOG_INFO or lower? The existing ReadKMsg=yes would then just be an alias for ReadKMsg=debug, i.e. read everything from LOG_DEBUG and lower.

filtering early by log level should be trivial. We'd still continue to ship ReadKMsg=yes upstream, but clearlinux could ship ReadKMsg=info. That means you can debug log as much as you want, but you'd also not lose really relevant kernel info either.

or short: you try to discern "boot" vs "runtime", and it appears to me given your usecase "debug" vs. "non-debug" would cover things better and also be easier to implement.

Does that make sense?

**Comment by poettering** [comment]
(if that was implemented a future further optimization might even be to add a new ioctl to /dev/kmsg which would allow userspace to specify the minimum log level of the msgs it wants to stream out of it. we could then tell the kernel and not even get woken up on debug msgs. such a kernel patch should be pretty straightforward.)

**Comment by ahkok** [comment]
On 4/7/20 6:57 AM, Lennart Poettering wrote:
> filtering early by log level should be trivial. We'd still continue to 
> ship ReadKMsg=yes upstream, but clearlinux could ship ReadKMsg=info. 
> That means you can debug log as much as you want, but you'd also not 
> lose really relevant kernel info either.


I like this, but, what we do is specifically set initcall_debug=1, which 
isn't really about debugging (systemd-bootchartd uses it). Maybe we just 
need to stop doing that entirely.

I can totally see value in changing the level - however - I would 
totally want this to be a runtime changeable value that you can tune 
with `journalctl` as a minimum.

**Comment by poettering** [comment]
> I like this, but, what we do is specifically set initcall_debug=1, which isn't really about debugging (systemd-bootchartd uses it).

But the additional output initcall_debug=1 generates is at LOG_DEBUG log level, no? so this is nicely covered, no?

> Maybe we just need to stop doing that entirely. I can totally see value in changing the level - however - I would totally want this to be a runtime changeable value that you can tune with `journalctl` as a minimum.

You can always change journald.conf and restart journald, no?

**Comment by ahkok** [comment]
Sorry for delay - This sounds like a really good solution here and I think it would work extremely well for ClearLinux out of the box too. I would argue that by default most distributions wouldn't care too much about debug level output and put INFO and lower priority in the journal, and that's the intended / desired thing we want both at boot time and later on.

I'll see whether I can pull this off myself (at least for an initial revision) since I don't think it'll be very complex.

**Comment by poettering** [comment]
OK, then let's close this here. Would be happy to review & merge a patch that turns ReadKMsg= into something that takes a log level!

**Pull request closed without merging**